### PR TITLE
More explainable recipes

### PR DESCRIPTION
**Pull request by kmontgom2400**
This adds the three explainable recipes:

model_skopes_rules.py : Skopes decision rules
model_decision_tree_linear_combo.py:  Runs a decision tree, then runs a linear or logistic model on each leaf node
model_gam.py: Generalized Additive Models

### CODE REVIEW

File: models/mli/model_gam.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+    def fit(self, X, y, sample_weight=None, eval_set=None, sample_weight_eval_set=None, **kwargs):
+        
+        orig_cols = list(X.names)
+       
+        import pandas as pd
+        import numpy as np
+        from sklearn.preprocessing import OneHotEncoder
+        from collections import Counter
+        from pygam import LinearGAM, LogisticGAM
```
Comment (navdeep-G) [inline models/mli/model_gam.py:88]: H2O-3 now has GAMs. Might be worth looking into.

File: models/mli/model_skopes_rules.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+            n_estimators= [10]
+            precision_min = [0.1, 0.2]
+            recall_min=[0.01]
+            max_samples = [0.8, 1.0]
+            max_samples_features = [0.8, 1.0]
+            max_depth = [3, 4]
+            max_features = ["auto"]
+            min_samples_split = [2]
+            bootstrap= [True, False]     
+            bootstrap_features = [True, False]
```
Comment (navdeep-G) [inline models/mli/model_skopes_rules.py:72]: Is there reasoning behind parameters selected for these models? Maybe good to comment.

File: models/mli/model_skopes_rules.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+              
+        # Get the logger if it exists
+        logger = None
+        if self.context and self.context.experiment_id:
+            logger = make_experiment_logger(experiment_id=self.context.experiment_id,
+                                                 tmp_dir=self.context.tmp_dir,
+                                                 experiment_tmp_dir=self.context.experiment_tmp_dir)
+            
+        # Set up temp folder
+        tmp_folder = self._create_tmp_folder(logger)
```
Comment (navdeep-G) [inline models/mli/model_skopes_rules.py:136]: I don't see a use for this. Do we save anything?

File: models/mli/model_gam.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+            max_iter = [100, 1000]
+
+        elif accuracy >= 5:
+            lam = [0,  0.01, 0.1, 1.0, 10.0]
+            max_iter = [100]
+
+        else:
+            lam = [0,  0.01, 0.1, 1.0, 10.0]
+            max_iter = [100]
+
```
Comment (navdeep-G) [inline models/mli/model_gam.py:47]: Is there reasoning behind parameters selected for these models? Maybe good to comment.

File: models/mli/model_decision_tree_linear_combo.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+    def mutate_params(self, accuracy=10, **kwargs):
+        if accuracy > 8:
+            tree_depth = [4, 5]
+        elif accuracy >= 5:
+            tree_depth = [3, 4]
+        else:
+            tree_depth = [2]
+
+        self.params["tree_depth"] = np.random.choice(tree_depth)
```
Comment (navdeep-G) [inline models/mli/model_decision_tree_linear_combo.py:43]: Is there reasoning behind parameters selected for these models? Maybe good to comment.

File: models/mli/model_decision_tree_linear_combo.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
+"""Decision tree plus linear model"""
```
[the remark this answers is not in the document]
Comment (kmontgom2400) [reply models/mli/model_decision_tree_linear_combo.py:1]: Yes

File: models/mli/model_gam.py (hunk 1 above)
Comment (kmontgom2400) [reply models/mli/model_gam.py:88]: Looking into it...

File: models/mli/model_skopes_rules.py (hunk 3 above)
Comment (kmontgom2400) [reply models/mli/model_skopes_rules.py:136]: Yes, it's saving the skope rules to a csv file.

**Review by navdeep-G — APPROVED** [review]

**Pull request merged**